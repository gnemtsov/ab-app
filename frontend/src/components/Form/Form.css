/**********Form**********/

.Form {
    width: 100%;
    text-align: left;
}

.Form * {
    box-sizing: border-box;
}

.Form.Horizontal {
    display: grid;
    grid-template-columns: auto min-content;
    grid-auto-rows: min-content;
    grid-column-gap: 0.5rem;
    grid-row-gap: 0.4rem;
}

@media (min-width: 40rem) {
    .Form.Horizontal {
        grid-template-columns: minmax(min-content, max-content) minmax(min-content, auto) min-content;
        grid-row-gap: 0;
    }
}

.Form.Inline {
    display: flex;
    flex-wrap: wrap;
    align-items: flex-start;
}

.Form.Inline>.FormElement {
    display: flex;
    margin-right: 1.5rem;
    align-items: flex-start;
}

/**********Info icons**********/

.DefaultInfoIcon {
    height: 100%;
    background-position: center;
    background-repeat: no-repeat;
    background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0IiBmaWxsPSJub25lIiBzdHJva2U9IiM2NjY2NjYiIHN0cm9rZS13aWR0aD0iMiIgc3Ryb2tlLWxpbmVjYXA9InJvdW5kIiBzdHJva2UtbGluZWpvaW49InJvdW5kIj48Y2lyY2xlIGN4PSIxMiIgY3k9IjEyIiByPSIxMCI+PC9jaXJjbGU+PGxpbmUgeDE9IjEyIiB5MT0iMTYiIHgyPSIxMiIgeTI9IjEyIj48L2xpbmU+PGxpbmUgeDE9IjEyIiB5MT0iOCIgeDI9IjEyIiB5Mj0iOCI+PC9saW5lPjwvc3ZnPg==);
}

/**********Tooltips**********/

.Tooltip {
    max-width: 24rem;
    background-color: #ccc !important;
    color: #222 !important;
    padding: 0.5rem 0.8rem !important;
    font-size: 0.8rem !important;
    line-height: 1rem !important;
}

.Tooltip::after {
    border-top-color: transparent !important;
    border-right-color: transparent !important;
    border-bottom-color: transparent !important;
    border-left-color: transparent !important;
}

/**********Submit**********/

.Form.Horizontal .SubmitButton {
    justify-self: start;
    grid-column: 1 / 2;
}

@media (min-width: 40rem) {
    .Form.Horizontal .SubmitButton {
        grid-column: 2 / 3;
    }
}

.DoneText {
    color: green;
    margin-left: 1rem;
    font-weight: bold;
    opacity: 0;
    transition: opacity 0.5s linear;
}

.DoneText.Visible {
    opacity: 1;
}