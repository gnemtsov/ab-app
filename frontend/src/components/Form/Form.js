import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import ReactTooltip from 'react-tooltip';

import Button from '../UI/Button/Button';
import FormElement from './FormElement/FormElement';
import classes from './Form.css';

export default class AbForm extends Component {

    formClasses = [classes.Form]

    static defaultProps = {
        infoIcon: <div className={classes.DefaultInfoIcon}></div>,
        layout: 'horizontal',
        doneText: 'Done!',
        className: '',
        doneTextDuration: 2000,
        buttonText: ['Submit', 'Sending...']
    }

    static propTypes = {
        infoIcon: PropTypes.node,
        layout: PropTypes.oneOf(['horizontal', 'inline']),
        doneText: PropTypes.string,
        className: PropTypes.string,
        submitHandler: PropTypes.func.isRequired,
        doneTextDuration: PropTypes.number,
        buttonText: PropTypes.array,
        fields: PropTypes.array.isRequired
    }

    state = {
        sending: false,
        saved: false,
        fields: []
    }

    constructor(props) {
        super(props);

        const { layout, className, fields } = props;
        this.formClasses.push(layout === 'inline' ? classes.Inline : classes.Horizontal);
        this.formClasses.push(className);

        this.state.fields = fields.map((field) => { //-------> setup form fields in state
            this.setFieldDefaultValue(field, field.value);
            field.message = '';

            if (field.validators === undefined) {
                field.validators = [];
            } else {
                field.validators = field.validators.map((validator) => {
                    let f =
                        typeof validator.f === 'function' ?
                            validator.f :
                            // eslint-disable-next-line no-new-func
                            new Function("return " + validator.f)();                   
                    f.params = validator.params || [];
                    f.message = validator.message || '';
                    if(f.params.length){
                        f.message = f.message.replace(/%([0-9]+)%/g, (...args) => f.params[Number(args[1])]);
                    }
                    return f;
                });
            }

            return field;
        });
    }

    isFormReady = () => {
        for (const { required, value, message } of this.state.fields) {
            if ((required && value === '') || message !== '') {
                return false;
            }
        }
        return true;
    }

    setFieldDefaultValue = (field, value) => {
        const shouldClear = value === undefined || value === null;
        switch (field.type) {
            case 'Boolean':
                field.value = shouldClear ? false : value;
                break;
            case 'Date':
                field.value = shouldClear ? moment() : moment(value);
                break;
            default:
                field.value = shouldClear ? '' : value;
        }
        return field;
    }

    inputChangedHandler = (data, name) => {
        const i = this.state.fields.findIndex(field => name === field.name);
        const { type, validators } = this.state.fields[i];

        let stateValue;
        let validateValue;
        switch (type) {
            case 'Boolean':
                stateValue = data.target.checked;
                validateValue = data.target.checked;
                break;
            case 'Date':
                stateValue = data;
                validateValue = data.format('YYYY-MM-DD').toString();
                break;
            default:
                stateValue = data.target.value;
                validateValue = data.target.value;
        }

        let message = '';
        for (const f of validators) {
            if (!f(validateValue, ...f.params)) {
                message = f.message;
                break;
            }
        }

        let field = {
            ...this.state.fields[i],
            message: message,
            value: stateValue
        }

        this.setState((prevState, props) => {
            prevState.fields[i] = field;
            return { fields: prevState.fields };
        })
    }

    submitHandler = (event) => {
        event.preventDefault();

        this.setState({ sending: true });

        const values = {};
        for (const { type, name, value } of this.state.fields) {
            switch (type) {
                case 'Date':
                    values[name] = value.format('YYYY-MM-DD').toString();
                    break;
                default:
                    values[name] = value;
            }
        }

        this.props.submitHandler(values)
            .catch(error => {
                const errorField = error.response.data.field;
                this.setState(prevState => {
                    const i = prevState.fields.findIndex(field => errorField.name === field.name);
                    prevState.fields[i] = {
                        ...prevState.fields[i],
                        message: errorField.message
                    };
                    return {
                        sending: false,
                        fields: prevState.fields
                    };
                })
            })
            .then(result => {
                if (typeof result === 'function') {
                    return result();
                } else {
                    this.setState(prevState => {
                        let fields = prevState.fields;
                        if (result === true) { //clear form fields
                            fields = fields.map(field => this.setFieldDefaultValue(field, null));
                        }
                        return {
                            saved: true,
                            sending: false,
                            fields: fields
                        };
                    });
                    this.timerId = setTimeout(
                        () => this.setState({ saved: false }),
                        this.props.doneTextDuration
                    );
                }
            });
    }

    componentWillUnmount() {
        clearTimeout(this.timerId);
    }

    render() {
        const {
            infoIcon,
            layout,
            doneText,
            buttonText
        } = this.props;

        //form elements
        let formElements = [];
        for (const field of this.state.fields) {
            let element =
                <FormElement
                    key={field.name}
                    id={field.name}
                    infoIcon={infoIcon}
                    layout={layout}
                    {...field}
                    inputChanged={(data) => this.inputChangedHandler(data, field.name)} />

            if (layout === 'inline') {
                element =
                    <div
                        className={classes.FormElement}>
                        {element}
                    </div>;
            }

            formElements.push(element);
        }

        //form button
        const doneTextClasses = [classes.DoneText];
        if (this.state.saved) {
            doneTextClasses.push(classes.Visible);
        }
        let button =
            <div
                key='SubmitButton'
                className={classes.SubmitButton}>
                <Button
                    type="Submit"
                    disabled={this.state.sending || !this.isFormReady()}>
                    {this.state.sending ? buttonText[1] : buttonText[0]}
                </Button>
                <span className={doneTextClasses.join(' ')}>{doneText}</span>
            </div>;

        return ([
            <form
                key="AbForm"
                className={this.formClasses.join(' ')}
                onSubmit={this.submitHandler}>
                {formElements}
                {button}
            </form >,
            <ReactTooltip
                key="ReactTooltip"
                type="info"
                place="right"
                className={classes.Tooltip} />
        ]);
    }
}
