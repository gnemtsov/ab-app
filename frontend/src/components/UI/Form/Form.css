.Form input[type="text"],
.Form input[type="password"],
.Form select,
.Form textarea {
    font-family: 'Arimo', sans-serif;
}
