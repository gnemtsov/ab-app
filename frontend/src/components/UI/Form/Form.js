import React, { Component } from 'react';

import axios from 'axios';

import AbForm from 'react-ab-form';

import Spinner from '../Spinner/Spinner';
import Icon from '../Icon/Icon';
import classes from './Form.css';

export class Form extends Component {
    conf = {
        className: classes.Form,
        infoIcon: <Icon name="info" width="26" height="26" stroke="#666666" />
    }

    state = {}

    constructor(props) {
        super(props);

        let { conf, fields } = props;

        this.conf = {
            ...this.conf,
            ...conf,
            submitHandler: this.submitHandler
        }

        this.state = { fields }
    }

    componentDidMount() {
        if (this.state.fields === undefined) {
            axios.get(this.props.api)
                .then(result => this.setState({ fields: result.data }));
        }
    }

    submitHandler = (values) => {
        //if the server responds with error, it will be catched by AbForm component
        return axios.post(this.props.api, values)
            .then(response => {
                if (typeof this.props.submitted === 'function') {
                    return () => this.props.submitted(response.data); 
                } else {
                    return this.props.submitted; //don't clear form by default
                }
            });
    }

    render() {
        let form = <Spinner />;

        if (this.state.fields !== undefined) {
            form = <AbForm {...this.conf} {...this.state} />;
        }

        return (
            <React.Fragment>
                <h1>{this.props.title}</h1>
                {form}
            </React.Fragment>
        );
    }
}

export default Form;