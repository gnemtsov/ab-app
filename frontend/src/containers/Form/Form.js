import React, { Component } from 'react';
import axios from 'axios';

import ErrorBoundary from '../../hoc/errorBoundary/errorBoundary';
import Spinner from '../../components/UI/Spinner/Spinner';
import Icon from '../../components/UI/Icon/Icon';
import FormComponent from '../../components/Form/Form';

import classes from './Form.css';

export class Form extends Component {
    conf = {
        className: classes.Form,
        infoIcon: <Icon name="info" width="26" height="26" stroke="#666666" />
    }

    state = {}

    constructor(props) {
        super(props);

        let { conf, fields } = props;

        this.conf = {
            ...this.conf,
            ...conf,
            submitHandler: this.submitHandler
        }

        this.state = { fields }
    }

    componentDidMount() {
        if (this.state.fields === undefined) {
            axios.get(this.props.api)
                .then(result => this.setState({ fields: result.data }));
        }
    }

    submitHandler = (values) => {
        //if the server responds with error, it will be catched by AbForm component
        return axios.post(this.props.api, values)
            .then(response => {
                if (typeof this.props.submitted === 'function') {
                    return () => this.props.submitted(response.data);
                } else {
                    return this.props.submitted; //don't clear form by default
                }
            });
    }

    render() {
        let form = <Spinner />;

        if (this.state.fields !== undefined) {
            form = <FormComponent {...this.conf} {...this.state} />;
        }

        return (
            <React.Fragment>
                <h1>{this.props.title}</h1>
                <ErrorBoundary>{form}</ErrorBoundary>
            </React.Fragment>
        );
    }
}

export default Form;