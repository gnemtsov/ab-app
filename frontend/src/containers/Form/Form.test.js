import React from 'react';

import { configure, render, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Form from './Form';

configure({ adapter: new Adapter() });

describe('Form', () => {
	it('Renders correctly before loading', () => {
		const conf = {
			buttonText: ['Add', 'Adding..'],
			doneText: 'Department was added!'
		};
        const wrapper = shallow (
            <Form
                conf={conf}
                title="New department"
                api="/departments/add" />
        );
		expect(wrapper).toMatchSnapshot();
	});
	it('Renders correctly after loading', () => {
		const conf = {
			buttonText: ['Add', 'Adding..'],
			doneText: 'Department was added!'
		};
        const wrapper = shallow (
            <Form
                conf={conf}
                title="New department"
                api="/departments/add" />
        );
		return Promise.resolve(wrapper)
			.then(
				() => {
					wrapper.update();
					expect(wrapper).toMatchSnapshot();
				}
			);
	});
});

