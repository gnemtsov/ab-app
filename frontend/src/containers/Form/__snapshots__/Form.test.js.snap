// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Renders correctly after loading 1`] = `
<Fragment>
  <h1>
    New department
  </h1>
  <ErrorBoundary>
    <AbForm
      buttonText={
        Array [
          "Add",
          "Adding..",
        ]
      }
      className="Form"
      doneText="Department was added!"
      doneTextDuration={2000}
      fields={
        Array [
          Object {
            "allowedValues": null,
            "name": "d_id",
            "required": false,
            "table": "departments",
            "type": "Hidden",
            "validators": Array [
              Object {
                "f": "value => /^-?[0-9]*$/.test(value)",
                "message": "Value must a number",
                "name": "numType",
              },
              Object {
                "f": "(value, min) => value >= min",
                "message": "Value must be greater than %0%",
                "name": "numMin",
                "params": Array [
                  -2147483648,
                ],
              },
              Object {
                "f": "(value, max) => value <= max",
                "message": "Value must be smaller than %0%",
                "name": "numMax",
                "params": Array [
                  2147483647,
                ],
              },
            ],
          },
          Object {
            "allowedValues": null,
            "label": "Title",
            "name": "d_title",
            "placeholder": "Department title",
            "required": true,
            "table": "departments",
            "type": "String",
            "validators": Array [
              Object {
                "f": "(value, max) => value.length <= max",
                "message": "String must be shorter or equal to %0% characters",
                "name": "strMax",
                "params": Array [
                  128,
                ],
              },
              Object {
                "f": "(value, min, max) => value.length >= min && value.length <= max",
                "message": "Title must be bigger than %0% and smaller than %1% chars",
                "name": "strMinMax",
                "params": Array [
                  6,
                  64,
                ],
              },
            ],
            "value": "Test title",
          },
          Object {
            "allowedValues": null,
            "label": "Head",
            "name": "d_head",
            "placeholder": "Department head's name",
            "required": true,
            "table": "departments",
            "type": "String",
            "validators": Array [
              Object {
                "f": "(value, max) => value.length <= max",
                "message": "String must be shorter or equal to %0% characters",
                "name": "strMax",
                "params": Array [
                  128,
                ],
              },
              Object {
                "f": "(value, min, max) => value.length >= min && value.length <= max",
                "message": "Head must be bigger than %0% and smaller than %1% chars",
                "name": "strMinMax",
                "params": Array [
                  3,
                  32,
                ],
              },
            ],
            "value": "Test head",
          },
          Object {
            "allowedValues": null,
            "label": "Size",
            "name": "d_size",
            "placeholder": "Department size",
            "required": true,
            "table": "departments",
            "type": "Number",
            "validators": Array [
              Object {
                "f": "value => /^-?[0-9]*$/.test(value)",
                "message": "Value must a number",
                "name": "numType",
              },
              Object {
                "f": "(value, min) => value >= min",
                "message": "Value must be greater than %0%",
                "name": "numMin",
                "params": Array [
                  0,
                ],
              },
              Object {
                "f": "(value, max) => value <= max",
                "message": "Value must be smaller than %0%",
                "name": "numMax",
                "params": Array [
                  255,
                ],
              },
              Object {
                "f": "(value, min) => value >= min",
                "message": "There must be at least %0% computers",
                "name": "numMin",
                "params": Array [
                  5,
                ],
              },
            ],
            "value": "13",
          },
          Object {
            "allowedValues": null,
            "label": "Created",
            "name": "d_created",
            "placeholder": "Department creation date",
            "required": true,
            "table": "departments",
            "type": "Date",
            "validators": Array [
              Object {
                "f": "value => /^[0-9]{4}-[0-9]{1,2}-[0-9]{1,2}$/.test(value)",
                "message": "Value must be a date",
                "name": "strIsDate",
              },
              Object {
                "f": "(value, min) => new Date(value) >= new Date(min)",
                "message": "Date must be bigger than %0%",
                "name": "dateMin",
                "params": Array [
                  "1000-01-01",
                ],
              },
              Object {
                "f": "(value, max) => new Date(value) <= new Date(max)",
                "message": "Date must be less than %0%",
                "name": "dateMax",
                "params": Array [
                  "9999-12-31",
                ],
              },
            ],
            "value": "2018-05-01",
          },
        ]
      }
      infoIcon={
        <Icon
          height="26"
          name="info"
          stroke="#666666"
          width="26"
        />
      }
      layout="horizontal"
      submitHandler={[Function]}
    />
  </ErrorBoundary>
</Fragment>
`;

exports[`Form Renders correctly before loading 1`] = `
<Fragment>
  <h1>
    New department
  </h1>
  <ErrorBoundary>
    <spinner />
  </ErrorBoundary>
</Fragment>
`;
